packages/firestore/test/integration/api/bundle.test.ts

```diff
@@ -51,74 +45,133 @@ function verifyInProgress(
+);
+ */
+
+const TEMPLATE_PROJECT_ID = 'test-project';
+const BUNDLE_TEMPLATE =
+  '125{"metadata":{"id":"test-bundle","createTime":{"seconds":1001,"nanos":9999},"version":1,"totalDocuments":2,"totalBytes":1503}}374{"namedQuery":{"name":"limit","readTime":{"seconds":1000,"nanos":9999},"bundledQuery":{"parent":"projects/test-project/databases/(default)/documents","structuredQuery":{"from":[{"collectionId":"coll-1"}],"orderBy":[{"field":{"fieldPath":"bar"},"direction":"DESCENDING"},{"field":{"fieldPath":"__name__"},"direction":"DESCENDING"}],"limit":{"value":1}},"limitType":"FIRST"}}}381{"namedQuery":{"name":"limit-to-last","readTime":{"seconds":1000,"nanos":9999},"bundledQuery":{"parent":"projects/test-project/databases/(default)/documents","structuredQuery":{"from":[{"collectionId":"coll-1"}],"orderBy":[{"field":{"fieldPath":"bar"},"direction":"DESCENDING"},{"field":{"fieldPath":"__name__"},"direction":"DESCENDING"}],"limit":{"value":1}},"limitType":"LAST"}}}147{"documentMetadata":{"name":"projects/test-project/databases/(default)/documents/coll-1/a","readTime":{"seconds":1000,"nanos":9999},"exists":true}}218{"document":{"name":"projects/test-project/databases/(default)/documents/coll-1/a","createTime":{"seconds":1,"nanos":9},"updateTime":{"seconds":1,"nanos":9},"fields":{"k":{"stringValue":"a"},"bar":{"integerValue":1}}}}147{"documentMetadata":{"name":"projects/test-project/databases/(default)/documents/coll-1/b","readTime":{"seconds":1000,"nanos":9999},"exists":true}}218{"document":{"name":"projects/test-project/databases/(default)/documents/coll-1/b","createTime":{"seconds":1,"nanos":9},"updateTime":{"seconds":1,"nanos":9},"fields":{"k":{"stringValue":"b"},"bar":{"integerValue":2}}}}';
+
+apiDescribe('Bundles', (persistence: boolean) => {
   function verifySnapEqualTestDocs(snap: firestore.QuerySnapshot): void {
     expect(toDataArray(snap)).to.deep.equal([
       { k: 'a', bar: 1 },
       { k: 'b', bar: 2 }
     ]);
   }
 
+  /**
+   * Returns a valid bundle string from replacing project id in `BUNDLE_TEMPLATE` with the given
+   * db project id (also recalculate length prefixes).
+   */
+  function bundleString(db: firestore.FirebaseFirestore): string {
+    const projectId: string = db.app.options.projectId;
+
+    // Extract elements from BUNDLE_TEMPLATE, and replace project ID.
+    let runningElement: string = '';
+    const elements: string[] = [];
+    let depth = 0;
+    for (const char of BUNDLE_TEMPLATE) {
+      if (char === '{') {
+        runningElement += char;
+        depth += 1;
+      } else if (depth > 0) {
+        runningElement += char;
+      }
+
+      if (char === '}') {
+        depth -= 1;
+        if (depth === 0) {
+          // Obviously assuming TEMPLATE_PROJECT_ID only appear as project id in
+          // the bundle, not as document id or filed value.
+          elements.push(runningElement.replace(TEMPLATE_PROJECT_ID, projectId));
+          runningElement = '';
+        }
+      }
+    }
```

Comment:
This mostly performs the same processing on the same static string. Have you considered storing the bundle as an array of template strings?

```
const elements = [
     "abc/{0}/def",
      ...
].map(val => val.format(projectID));
```

You can then calculate the lengths as you are doing right now.

Reply:
Manually, yes.

Reply:
Done. Only `String.format` does not seem to exist, and `replace` should be enough for our purpose here?